File: CSES/Additional Problems/P2229.cpp
#include <bits/stdc++.h>

using namespace std;

long long MOD = 1000000007LL;

int main(){
   ios::sync_with_stdio(false);
   cin.tie(0);
   
   int n,m;
   cin >> n >> m;
   
   vector<vector<long long>> psum(2,vector<long long>(m+2,0LL));
   
   for(int k = 1; k <= m+1; k++){
      psum[0][k] = 1L;
   }
   
   for(int k = 1; k < n; k++){
      for(int j = 0; j <= m; j++){
         int r = j;
         int l = max(0,j-k);
         
         long long cur = psum[0][r+1]-psum[0][l];
         if(cur < 0) cur += MOD;
         psum[1][j+1] = psum[1][j] + cur;
         if(psum[1][j+1] >= MOD) psum[1][j+1] -= MOD;
      }
      
      swap(psum[0],psum[1]);
      //psum[0].swap(psum[1]);
   }
   
   long long answer = ((psum[0][m+1]-psum[0][m])%MOD + MOD)%MOD;
   cout << answer << endl;
   
   return 0;
}

File: CSES/Additional Problems/P2229.java
//make sure to make new file!
import java.io.*;
import java.util.*;

public class P2229{
   
   public static long MOD = 1000000007L;
   
   public static void main(String[] args)throws IOException{
      BufferedReader f = new BufferedReader(new InputStreamReader(System.in));
      PrintWriter out = new PrintWriter(System.out);
      
      StringTokenizer st = new StringTokenizer(f.readLine());
      
      int n = Integer.parseInt(st.nextToken());
      int m = Integer.parseInt(st.nextToken());
      
      long[][] psum = new long[2][m+2];
      for(int k = 1; k <= m+1; k++){
         psum[1][k] = 1L;
      }
      
      for(int k = 1; k < n; k++){
         for(int j = 0; j <= m; j++){
            //if(dp[k][j] == 0) continue;
            int i = k&1;
            int i2 = i^1;
            
            int r = j;
            int l = Math.max(0,j-k);
            
            long cur = psum[i][r+1] - psum[i][l];
            if(cur < 0) cur += MOD;
            psum[i2][j+1] = psum[i2][j] + cur;
            if(psum[i2][j+1] >= MOD) psum[i2][j+1] -= MOD;
         }
      }
      
      long answer = ((psum[n&1][m+1] - psum[n&1][m])%MOD + MOD)%MOD;
      out.println(answer);
      
      
      
      
      
      
      
      
      out.close();
   }
   
      
}